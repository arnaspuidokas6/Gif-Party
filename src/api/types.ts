export interface FetchGifsRequest {
    query?: string;
    limit?: string;
}

// https://developers.giphy.com/docs/api/schema/#image-object
export interface IGifResponse {
    imageUrl: string;
    title: string;
    userImage?: string;
    importedAt?: string;
    displayName?: string;
}

interface IUrl {
    url: string;
}

interface IImage {
    fixed_height: IUrl;
}

interface IUser {
    avatar_url?: string;
    display_name?: string;
}

// https://developers.giphy.com/docs/api/schema/#gif-object
export interface IGif {
    title?: string;
    images?: IImage;
    user?: IUser;
    import_datetime?: string;
}
