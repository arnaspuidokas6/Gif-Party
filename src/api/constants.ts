export const API_KEY = '5vXtKWdBdVUrfD74YUXuNvFVQ8OanFKS';
export const DEFAULT_LIST_ITEMS = 12;
export const VALIDATION_PATTERN = '[^A-Za-z0-9]';
