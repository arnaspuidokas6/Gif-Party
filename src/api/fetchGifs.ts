import axios from 'axios';
import { API_KEY } from './constants';
import { FetchGifsRequest, IGif, IGifResponse } from './types';

const createGif = (data: IGif): IGifResponse => {
    return {
        title: data.title ?? '',
        imageUrl: data?.images?.fixed_height.url ?? '',
        userImage: data?.user?.avatar_url ?? '',
        displayName: data?.user?.display_name ?? '',
    };
};

export const fetchGifs = async ({ query = 'happy', limit = '12' }: FetchGifsRequest): Promise<IGifResponse[]> => {
    const res: {
        data: { data: IGif[] };
    } = await axios.get(`https://api.giphy.com/v1/gifs/search?q=${query}&limit=${limit}&api_key=${API_KEY}`);

    return res.data.data.map(createGif);
};
