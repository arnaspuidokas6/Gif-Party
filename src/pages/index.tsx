import React, { createContext, Dispatch, FC, SetStateAction, useContext } from 'react';
import { IGifResponse } from '../api/types';

export const EMPTY_GIFS_LIST = [];
export const IS_VALID = true;
export const DEFAULT_SELECTED_ITEM = {} as IGifResponse;
export const DEFAULT_OPEN_MODAL = false;
const EMPTY_VOID = () => {};

interface IGifsContext {
    gifsList: IGifResponse[];
    selectedItem: IGifResponse;
    setSelectedItem: (value: IGifResponse) => void;
    openModal: boolean;
    setOpenModal: Dispatch<SetStateAction<boolean>>;
}

const GifsListContext = createContext<IGifsContext>({
    gifsList: EMPTY_GIFS_LIST,
    selectedItem: DEFAULT_SELECTED_ITEM,
    setSelectedItem: EMPTY_VOID,
    openModal: DEFAULT_OPEN_MODAL,
    setOpenModal: EMPTY_VOID,
});

export const useGifsContext = () => {
    const context = useContext(GifsListContext);
    if (!context) {
        throw new Error("This component can't be used outside GifsList component");
    }
    return context;
};

export const GifsListPage: FC<IGifsContext> = ({
    gifsList,
    selectedItem,
    setSelectedItem,
    children,
    openModal,
    setOpenModal,
}) => {
    const context: IGifsContext = {
        gifsList,
        selectedItem,
        setSelectedItem,
        setOpenModal,
        openModal,
    };
    return <GifsListContext.Provider value={context}>{children}</GifsListContext.Provider>;
};
