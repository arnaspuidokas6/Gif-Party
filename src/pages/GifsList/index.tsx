import React, { FC, useState, useEffect } from 'react';
import { fetchGifs } from '../../api/fetchGifs';
import { IGifResponse } from '../../api/types';
import { Card } from '../../components/Card';
import { Heading } from '../../components/Heading';
import { NoGifs } from '../../components/NoGifs';
import { SearchBar } from '../../components/SearchBar';
import '../../tailwind.output.css';

export const GifsList: FC = () => {
    const [gifsList, setGifsList] = useState<IGifResponse[]>();
    const [searchValue, setSearchValue] = useState<string>('');
    const [searchError, setSearchError] = useState<boolean>(false);

    const handleSearchChange = (event: { target: { value: string } }) => {
        const typedValue = event.target.value;
        const regexpPattern = '[^A-Za-z0-9]';

        if (typedValue.match(regexpPattern)) {
            setSearchError(true);
        } else {
            setSearchError(false);
            setSearchValue(typedValue);
        }
    };

    useEffect(() => {
        fetchGifs({ query: searchValue }).then((newGifsList) => setGifsList(newGifsList));
    }, [searchValue]);

    return (
        <div className="container mx-auto">
            <Heading />
            <SearchBar setSearchValue={handleSearchChange} isValid={!searchError} />
            <>
                {gifsList?.length ? (
                    <div className="mt-20 grid grid-cols-1 gap-8 md:grid-cols-2 xl:grid-cols-4">
                        {gifsList?.map((props, index) => (
                            <Card {...props} key={index} />
                        ))}
                    </div>
                ) : (
                    <NoGifs />
                )}
            </>
        </div>
    );
};
