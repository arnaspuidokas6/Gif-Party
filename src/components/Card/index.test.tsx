import React, { useState } from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { DEFAULT_SELECTED_ITEM, EMPTY_GIFS_LIST, GifsListContext } from '../../pages';
import { IGifResponse } from '../../api/types';
import { Card } from '.';

const gifsList = EMPTY_GIFS_LIST as IGifResponse[];
const selectedItem = DEFAULT_SELECTED_ITEM;
const openModal = false;
const cardData = { imageUrl: "https://giphy.com/embed/cmxiR3UgXYTh5QKJA6", title: "randomTitle" };
describe('Card', () => {
    const setOpenModal = jest.fn().mockResolvedValueOnce(true);
    const setSelectedItem = jest.fn().mockResolvedValueOnce(true);

   it('should match snapshot', async () => {
        const { container } = render(
        <GifsListContext.Provider value={{ gifsList, openModal, selectedItem, setSelectedItem, setOpenModal }}>
            <Card {...cardData} /> 
        </GifsListContext.Provider>
        );
        expect(container).toMatchSnapshot();
    });

   it('should have correct call', async () => {
        render(
        <GifsListContext.Provider value={{ gifsList, openModal, selectedItem, setSelectedItem, setOpenModal }}>
            <Card {...cardData} /> 
        </GifsListContext.Provider>
        );
       
        expect(screen.getByText('randomTitle')).toBeVisible();
        userEvent.click(screen.getByText('Preview'));
        expect(setOpenModal).toHaveBeenCalledTimes(1);
        expect(setOpenModal).toHaveBeenCalledWith(true);
        expect(setSelectedItem).toHaveBeenCalledTimes(1);
        expect(setSelectedItem).toHaveBeenCalledWith(cardData);
    });

})
