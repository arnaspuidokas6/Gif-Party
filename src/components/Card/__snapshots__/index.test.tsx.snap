// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should match snapshot 1`] = `
<div>
  <div
    class="rounded-lg overflow-hidden"
  >
    <div
      class="relative overflow-hidden pb-60"
    >
      <img
        alt="randomTitle"
        class="absolute h-full w-full object-cover object-center"
        src="https://giphy.com/embed/cmxiR3UgXYTh5QKJA6"
      />
    </div>
    <div
      class="relative bg-shadowBlue"
    >
      <div
        class="py-10 px-8"
      >
        <h3
          class="text-2xl font-bold text-teaGreen"
        >
          randomTitle
        </h3>
        <div
          class="mt-10 flex justify-between items-center"
        >
          <button
            class="flex items-center relative bg-white text-cambridgeBlue mr-4 p-2 top-0 right-0 rounded-full focus:outline-teaGreen"
          >
            Preview
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
